File: app/src/main/java/com/zj/goodvideo/ui/home/FeedAdapter.java
package com.zj.goodvideo.ui.home;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;

import androidx.annotation.NonNull;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.ViewDataBinding;
import androidx.paging.PagingDataAdapter;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.zj.goodvideo.BR;
import com.zj.goodvideo.R;
import com.zj.goodvideo.databinding.LayoutFeedTypeImageBinding;
import com.zj.goodvideo.databinding.LayoutFeedTypeVideoBinding;
import com.zj.goodvideo.model.Feed;
import com.zj.goodvideo.view.ListPlayerView;
import com.zj.libcommon.AbsPagedListAdapter;

public class FeedAdapter extends PagingDataAdapter<Feed, FeedAdapter.ViewHolder> {
    protected Context mContext;
    protected String mCategory;

    public FeedAdapter(Context context, String category) {
        super(new DiffUtil.ItemCallback<Feed>() {
            @Override
            public boolean areItemsTheSame(@NonNull Feed oldItem, @NonNull Feed newItem) {
                return oldItem.id == newItem.id;
            }

            @Override
            public boolean areContentsTheSame(@NonNull Feed oldItem, @NonNull Feed newItem) {
                return oldItem.equals(newItem);
            }
        });

        LayoutInflater inflater = LayoutInflater.from(context);
        mContext = context;
        mCategory = category;
    }

    @Override
    public int getItemViewType(int position) {
        Feed item = getItem(position);
        if (item == null)
            return 0;
        if (item.itemType == Feed.TYPE_IMAGE_TEXT) {
            return R.layout.layout_feed_type_image;
        } else if (item.itemType == Feed.TYPE_VIDEO) {
            return R.layout.layout_feed_type_video;
        }
        return 0;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ViewDataBinding binding = DataBindingUtil.inflate(LayoutInflater.from(parent.getContext()), viewType, parent, false);
        return new ViewHolder(binding.getRoot(), binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.bindData(getItem(position));

    }

    public class ViewHolder extends RecyclerView.ViewHolder {

        public ViewDataBinding mBinding;
        public ListPlayerView listPlayerView;
        public ImageView feedImage;

        public ViewHolder(@NonNull View itemView, ViewDataBinding binding) {
            super(itemView);
            mBinding = binding;
        }

        public void bindData(Feed item) {
            //这里之所以手动绑定数据的原因是 图片 和视频区域都是需要计算的
            //而dataBinding的执行默认是延迟一帧的。
            //当列表上下滑动的时候 ，会明显的看到宽高尺寸不对称的问题

            mBinding.setVariable(BR.feed, item);
            mBinding.setVariable(BR.lifeCycleOwner, mContext);
            if (mBinding instanceof LayoutFeedTypeImageBinding) {
                LayoutFeedTypeImageBinding imageBinding = (LayoutFeedTypeImageBinding) mBinding;
                feedImage = imageBinding.feedImage;
                imageBinding.feedImage.bindData(item.width, item.height, 16, item.cover);
                //imageBinding.setFeed(item);
                //imageBinding.interactionBinding.setLifeCycleOwner((LifecycleOwner) mContext);
            } else if (mBinding instanceof LayoutFeedTypeVideoBinding) {
                LayoutFeedTypeVideoBinding videoBinding = (LayoutFeedTypeVideoBinding) mBinding;
                videoBinding.listPlayerView.bindData(mCategory, item.width, item.height, item.cover, item.url);
                listPlayerView = videoBinding.listPlayerView;
                //videoBinding.setFeed(item);
                //videoBinding.interactionBinding.setLifeCycleOwner((LifecycleOwner) mContext);
            }
        }

        public boolean isVideoItem() {
            return mBinding instanceof LayoutFeedTypeVideoBinding;
        }

        public ListPlayerView getListPlayerView() {
            return listPlayerView;
        }
    }
}

File: app/src/main/java/com/zj/goodvideo/ui/home/FeedAdapter.kt
package com.zj.goodvideo.ui.home

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.databinding.DataBindingUtil
import androidx.databinding.ViewDataBinding
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.zj.goodvideo.BR
import com.zj.goodvideo.R
import com.zj.goodvideo.databinding.LayoutFeedTypeImageBinding
import com.zj.goodvideo.databinding.LayoutFeedTypeVideoBinding
import com.zj.goodvideo.model.Feed
import com.zj.goodvideo.view.ListPlayerView
import com.zj.goodvideo.view.PPImageView
import com.zj.libcommon.extention.AbsPagedListAdapter

open class FeedAdapter(private val context: Context, val category: String) :
    AbsPagedListAdapter<Feed, FeedAdapter.ViewHolder>(object :
        DiffUtil.ItemCallback<Feed>() {
        override fun areItemsTheSame(oldItem: Feed, newItem: Feed): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: Feed, newItem: Feed): Boolean {
            return oldItem == newItem
        }
    }) {

    private val inflater = LayoutInflater.from(context)

    override fun getItemViewType(position: Int): Int {
        val feed = getItem(position)
        return when (feed?.itemType) {
            Feed.TYPE_IMAGE_TEXT -> {
                R.layout.layout_feed_type_image
            }
            Feed.TYPE_VIDEO -> {
                R.layout.layout_feed_type_video
            }
            else -> 0
        }
    }

    override fun onCreateViewHolder2(parent: ViewGroup, viewType: Int): ViewHolder {
        val binding = DataBindingUtil.inflate<ViewDataBinding>(inflater, viewType, parent, false)
        return ViewHolder(binding.root, binding)
    }

    override fun onBindViewHolder2(holder: ViewHolder, position: Int) {
        holder.bindData(getItem(position))

        holder.itemView.setOnClickListener {

        }
    }

    inner class ViewHolder(itemView: View, val binding: ViewDataBinding) :
        RecyclerView.ViewHolder(itemView) {

        private var listPlayerView: ListPlayerView? = null
        private var feedImage: PPImageView? = null

        fun bindData(item: Feed?) {
            item?.run {
                binding.setVariable(BR.feed, this)
                binding.setVariable(BR.lifeCycleOwner, context)
                if (binding is LayoutFeedTypeVideoBinding) {
                    val videoBinding = binding
                    videoBinding.listPlayerView.bindData(category, width, height, cover, url)
                    listPlayerView = videoBinding.listPlayerView
                } else if (binding is LayoutFeedTypeImageBinding) {
                    val imageBinding = binding
                    imageBinding.feedImage.bindData(width, height, 16, cover ?: "")
                    feedImage = imageBinding.feedImage
                }
            }
        }

        fun isVideoItem() = binding is LayoutFeedTypeVideoBinding

        fun getListPlayerView() = listPlayerView
    }

}